### PR TITLE
[SPARK-37620][PYTHON] Use more precise types for SparkContext Optional fields (i.e. _gateway, _jvm)

### PR DESCRIPTION
**Pull request by ByronHsu**
### What changes were proposed in this pull request?
In SPARK-37152, we agreed to temporarily hide numerous errors by a quick fix. However, we should use a more precise type (optional) to describe sparkcontext's internal field, including _gateway, _jvm, and _active_spark_context.

### Why are the changes needed?

Use the precise type for sparkcontext's field

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

regular test

### CODE REVIEW

**Comment by ByronHsu** [comment]
cc: @zero323 @xinrong-databricks

**Comment by ByronHsu** [comment]
@zero323 Should we use `cast` or `assert`? I think `assert` makes more sense because if that variable is None, some errors might occur.

**Comment by zero323** [comment]
> @zero323 Should we use `cast` or `assert`? I think `assert` makes more sense because for example, it might be problematic to `cast` a `None` to `SparkContext`.

Yup, let's go with `assert`.

File: python/pyspark/context.py [hunk 1]
```diff
@@ -152,10 +152,10 @@ class SparkContext(object):
     """
 
     # set assignment ignore temporarily to prevent errors from other files
```
Comment (itholic) [inline python/pyspark/context.py:154]: So, then I think we don't need this comment anymore ?

Comment (ByronHsu) [reply python/pyspark/context.py:154]: make sense. I will revise it.

File: python/pyspark/sql/functions.py [hunk 2]
```diff
@@ -1081,7 +1081,9 @@ def approx_count_distinct(col: "ColumnOrName", rsd: Optional[float] = None) -> C
     >>> df.agg(approx_count_distinct(df.age).alias('distinct_ages')).collect()
     [Row(distinct_ages=2)]
     """
-    sc = SparkContext._active_spark_context  # type: ignore[attr-defined]
+    sc = SparkContext._active_spark_context
+    assert sc is not None
+    assert sc._jvm is not None
```
Comment (zero323) [inline python/pyspark/sql/functions.py:1086]: Nit: (I don't have strong feelings about it)  Doing this in a single line should work just fine and be preferable in the future:

```python
    assert sc is not None and sc._jvm is not None
```

File: python/pyspark/util.py [hunk 3]
```diff
@@ -402,18 +401,16 @@ def start(self) -> None:
...
-            )
+            assert SparkContext._active_spark_context is not None
+            self._props = SparkContext._active_spark_context._jsc.sc().getLocalProperties().clone()
         return super(InheritableThread, self).start()
 
     @staticmethod
     def _clean_py4j_conn_for_current_thread() -> None:
         from pyspark import SparkContext
 
         jvm = SparkContext._jvm  # type: ignore[attr-defined]
```
Comment (zero323) [inline python/pyspark/util.py:412]: We shouldn't need this ignore anymore, should we?

**Comment by zero323** [comment]
I finished the first past and so far looks good ‒ thanks for working on this @ByronHsu!

**Comment by ByronHsu** [comment]
@zero323 I will look into SPARK-37154 after this ticket

**Comment by itholic** [comment]
Looks fine to me

File: python/pyspark/sql/column.py [hunk 4]
```diff
@@ -82,6 +84,7 @@ def _to_seq(
     """
     if converter:
         cols = [converter(c) for c in cols]
+    assert sc._jvm is not None
     return sc._jvm.PythonUtils.toSeq(cols)  # type: ignore[attr-defined]
```
Comment (zero323) [inline python/pyspark/sql/column.py:88]: We don't need this ignore, do we?

File: python/pyspark/sql/context.py [hunk 5]
```diff
@@ -168,6 +168,7 @@ def getOrCreate(cls: Type["SQLContext"], sc: SparkContext) -> "SQLContext":
             cls._instantiatedContext is None
             or SQLContext._instantiatedContext._sc._jsc is None  # type: ignore[union-attr]
         ):
+            assert sc._jvm is not None
             jsqlContext = (
                 sc._jvm.SparkSession.builder()  # type: ignore[attr-defined]
                 .sparkContext(sc._jsc.sc())  # type: ignore[attr-defined]
```
Comment (zero323) [inline python/pyspark/sql/context.py:174]: This two should be obsolete as well.

File: python/pyspark/sql/context.py [hunk 6]
```diff
@@ -733,6 +734,7 @@ def _createForTesting(cls, sparkContext: SparkContext) -> "HiveContext":
         confusing error messages.
         """
         jsc = sparkContext._jsc.sc()  # type: ignore[attr-defined]
+        assert sparkContext._jvm is not None
         jtestHive = sparkContext._jvm.org.apache.spark.sql.hive.test.TestHiveContext(  # type: ignore[attr-defined]
```
Comment (zero323) [inline python/pyspark/sql/context.py:738]: Ditto, probably obsolete.

File: python/pyspark/sql/observation.py [hunk 7]
```diff
@@ -103,7 +103,8 @@ def _on(self, df: DataFrame, *exprs: Column) -> DataFrame:
         assert self._jo is None, "an Observation can be used with a DataFrame only once"
 
         self._jvm = df._sc._jvm  # type: ignore[attr-defined]
```
Comment (zero323) [inline python/pyspark/sql/observation.py:105]: Ditto.

File: python/pyspark/sql/pandas/conversion.py [hunk 8]
```diff
@@ -617,9 +617,8 @@ def create_RDD_server():
         jrdd = self._sc._serialize_to_jvm(  # type: ignore[attr-defined]
```
Comment (zero323) [inline python/pyspark/sql/pandas/conversion.py:617]: Ditto.

File: python/pyspark/sql/readwriter.py [hunk 9]
```diff
@@ -180,6 +180,7 @@ def load(
         elif path is not None:
             if type(path) != list:
                 path = [path]  # type: ignore[list-item]
+            assert self._spark._sc._jvm is not None
             return self._df(
                 self._jreader.load(
                     self._spark._sc._jvm.PythonUtils.toSeq(path)  # type: ignore[attr-defined]
```
Comment (zero323) [inline python/pyspark/sql/readwriter.py:186]: Also this one.

File: python/pyspark/sql/readwriter.py [hunk 10]
```diff
@@ -424,6 +427,7 @@ def text(
 
         if isinstance(paths, str):
             paths = [paths]
+        assert self._spark._sc._jvm is not None
         return self._df(
             self._jreader.text(
                 self._spark._sc._jvm.PythonUtils.toSeq(paths)  # type: ignore[attr-defined]
```
Comment (zero323) [inline python/pyspark/sql/readwriter.py:433]: Ditto.

File: python/pyspark/sql/readwriter.py [hunk 11]
```diff
@@ -541,6 +545,7 @@ def csv(
         if isinstance(path, str):
             path = [path]
         if type(path) == list:
+            assert self._spark._sc._jvm is not None
             return self._df(
                 self._jreader.csv(
                     self._spark._sc._jvm.PythonUtils.toSeq(path)  # type: ignore[attr-defined]
```
Comment (zero323) [inline python/pyspark/sql/readwriter.py:551]: Ditto.

File: python/pyspark/sql/streaming.py [hunk 12]
```diff
@@ -1270,6 +1271,7 @@ def func_with_open_process_close(partition_id: Any, iterator: Iterator) -> Itera
 
         serializer = AutoBatchedSerializer(CPickleSerializer())
         wrapped_func = _wrap_function(self._spark._sc, func, serializer, serializer)
+        assert self._spark._sc._jvm is not None
         jForeachWriter = self._spark._sc._jvm.org.apache.spark.sql.execution.python.PythonForeachWriter(  # type: ignore[attr-defined]
```
Comment (zero323) [inline python/pyspark/sql/streaming.py:1275]: Another one.

File: python/pyspark/util.py [hunk 13]
```diff
@@ -326,16 +326,14 @@ def inheritable_thread_target(f: Callable) -> Callable:
...
-        )
+        assert SparkContext._active_spark_context is not None
+        properties = SparkContext._active_spark_context._jsc.sc().getLocalProperties().clone()
 
         @functools.wraps(f)
         def wrapped(*args: Any, **kwargs: Any) -> Any:
             try:
                 # Set local properties in child thread.
+                assert SparkContext._active_spark_context is not None
                 SparkContext._active_spark_context._jsc.sc().setLocalProperties(  # type: ignore[attr-defined]
```
Comment (zero323) [inline python/pyspark/util.py:337]: Also, this one.

File: python/pyspark/util.py [hunk 14]
```diff
@@ -379,6 +377,7 @@ def __init__(self, target: Callable, *args: Any, **kwargs: Any):
             def copy_local_properties(*a: Any, **k: Any) -> Any:
                 # self._props is set before starting the thread to match the behavior with JVM.
                 assert hasattr(self, "_props")
+                assert SparkContext._active_spark_context is not None
                 SparkContext._active_spark_context._jsc.sc().setLocalProperties(  # type: ignore[attr-defined]
```
Comment (zero323) [inline python/pyspark/util.py:381]: And the last one :)

**Review by zero323 — APPROVED** [review]
LGTM. There is a bunch of `ignores` that we can drop in this sweep (listed above), but it is OK, if we don't.

File: python/pyspark/context.py [hunk 15]
```diff
@@ -288,9 +287,11 @@ def _do_init(
 
         # Create a single Accumulator in Java that we'll send all our updates through;
         # they will be passed back to us through a TCP server
+        assert self._gateway is not None
```
Comment (srowen) [inline python/pyspark/context.py:290]: Why are the asserts needed, out of curiosity

Comment (zero323) [reply python/pyspark/context.py:290]: > Why are the asserts needed, out of curiosity

Long story short ‒ `_gateway`, `_jvm`, etc. are class variables that might have or might have not been initialized 

https://github.com/apache/spark/blob/6e45b04db48008fa033b09df983d3bd1c4f790ea/python/pyspark/context.py#L155-L158

so all are defined `Optional[_]`.

On runtime we blindly assume that we deal with active context and these are set, but that's not clear for the type checker ‒ these assertions serve as `mypy` hints "hey, I know this stuff is not `None` here".

There are other ways of handling this (`casts`, `ignores`, we even discussed helper methods, but this seems to be the least intrusive approach and the closest to our actual expectations).

**Comment by xinrong-databricks** [comment]
Scala 2.13 build with SBT failed, which may not be relevant to the PR.

LGTM, thanks!

**Comment by zero323** [comment]
Merged into master. 

Thanks all!

**Pull request closed without merging**